File: 615 Is It A Tree?/solution.c
// #解法一
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#define MAX_NUM 50000

int root[MAX_NUM + 1];
bool isTree;

void init() {
    memset(root, 0, sizeof(root));
    isTree = true;
}

int findRoot(int child) {
    return root[child] == 0 || root[child] == child ? (root[child] = child) : (root[child] = findRoot(root[child]));
}

int main() {
    init();
    int cases = 0, parent, child;
    while(scanf("%d %d", &parent, &child) && parent >= 0) {
        if(parent == 0 && child == 0) {
            if(!isTree)
                printf("Case %d is not a tree.\n", ++cases);
            else {
                int rootCount = 0;
                for(int n = 1; n <= MAX_NUM && rootCount <= 1; ++n) {
                    if(root[n] != 0 && n == findRoot(n))
                        ++rootCount;
                }
                printf("Case %d is %sa tree.\n", ++cases, rootCount <= 1 ? "" : "not ");
            }
            init();
        } else {
            int rootParent = findRoot(parent), rootChild = findRoot(child);
            if(rootParent != rootChild)
                root[rootChild] = rootParent;
            else
                isTree = false;
        }
    }
    return 0;
}

// #解法二
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define MAX_NODES 50000

int **conn, connCount[MAX_NODES + 1], parentCount[MAX_NODES + 1], edgeCount, startVertex;
bool visited[MAX_NODES + 1], isTree = true;

void init() {
    memset(connCount + 1, 0, MAX_NODES * sizeof(int));
    memset(parentCount + 1, 0, MAX_NODES * sizeof(int));
    edgeCount = startVertex = 0;
    isTree = true;
}

void dfs(int vertex, int *nodeCount) {
    visited[vertex] = true;
    ++(*nodeCount);
    for(int i = 0; i < connCount[vertex]; ++i) {
        if(!visited[conn[vertex][i]])
            dfs(conn[vertex][i], nodeCount);
    }
}

int main() {
    conn = (int**)calloc(MAX_NODES + 1, sizeof(int*));
    int cases = 0, parent, child;
    while(scanf("%d %d", &parent, &child) && parent >= 0) {
        if(parent == 0 && child == 0) {
            if(!isTree)
                printf("Case %d is not a tree.\n", ++cases);
            else if(edgeCount == 0)
                printf("Case %d is a tree.\n", ++cases);
            else {
                memset(visited + 1, 0, MAX_NODES * sizeof(bool));
                int nodeCount = 0;
                dfs(startVertex, &nodeCount);
                if(nodeCount == edgeCount + 1)
                    printf("Case %d is a tree.\n", ++cases);
                else
                    printf("Case %d is not a tree.\n", ++cases);
            }
            init();
        } else {
            if(parentCount[child] < 1) {
                conn[parent] = (int*)realloc(conn[parent], (connCount[parent] + 1) * sizeof(int));
                conn[parent][connCount[parent]++] = child;
                conn[child] = (int*)realloc(conn[child], (connCount[child] + 1) * sizeof(int));
                conn[child][connCount[child]++] = parent;
                ++parentCount[child];
                ++edgeCount;
                startVertex = parent;
            } else
                isTree = false;
        }
    }
    for(int i = 0; i <= MAX_NODES; ++i)
        free(conn[i]);
    free(conn);
	return 0;
}

File: 615 Is It A Tree?/solution.java
// #解法一
import java.io.*;
import java.util.*;

public class Main {
    static int MAX_NUM = 50000;
    static int[] root = new int[MAX_NUM + 1];
    static ArrayList<Integer> nodes;
    static boolean isTree;
	public static void main(String[] args) throws IOException {
	    init();
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		int cases = 0;
		StringBuilder output = new StringBuilder();
		while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
    		int parent = (int)st.nval;
    		st.nextToken();
    		int child = (int)st.nval;
    		if(parent < 0)
    		    break;
    		if(parent == 0 && child == 0) {
                if(!isTree)
                    output.append("Case ").append(++cases).append(" is not a tree.\n");
                else {
                    int rootCount = 0;
                    for(int i = 0; i < nodes.size() && rootCount <= 1; ++i) {
                        if(nodes.get(i) == find(nodes.get(i)))
                            ++rootCount;
                    }
                    output.append("Case ").append(++cases).append(rootCount <= 1 ? " is a tree.\n" : " is not a tree.\n");
                }
                init();
            } else {
                int rootParent = find(parent), rootChild = find(child);
                if(rootParent != rootChild)
                    root[rootChild] = rootParent;
                else
                    isTree = false;
            }
		}
		System.out.print(output);
	}

    static void init() {
        Arrays.fill(root, 0);
        nodes = new ArrayList<>();
        isTree = true;
    }
    
    static int find(int x) {
        if(root[x] == 0) {
            nodes.add(x);
            return root[x] = x;
        }
    	return root[x] == x ? x : (root[x] = find(root[x]));
    }
}

// #解法二
import java.io.*;
import java.util.*;

public class Main {
    static HashMap<Integer, ArrayList<Integer>> conn;
    static HashMap<Integer, Integer> parentCount;
    static int edgeCount, startVertex;
    static boolean isTree;
	public static void main(String[] args) throws IOException {
		init();
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        int cases = 0, parent, child;
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (parent = (int)st.nval) >= 0 && st.nextToken() == StreamTokenizer.TT_NUMBER && (child = (int)st.nval) >= 0) {
            if(parent == 0 && child == 0) {
                if(!isTree)
                    output.append("Case ").append(++cases).append(" is not a tree.\n");
                else if(edgeCount == 0)
                    output.append("Case ").append(++cases).append(" is a tree.\n");
                else {
                    HashMap<Integer, Boolean> visited = new HashMap<>();
                    int[] nodeCount = new int[1];
                    dfs(visited, startVertex, nodeCount);
                    if(nodeCount[0] == edgeCount + 1)
                        output.append("Case ").append(++cases).append(" is a tree.\n");
                    else
                        output.append("Case ").append(++cases).append(" is not a tree.\n");
                }
                init();
            } else {
                if(conn.containsKey(parent))
                    conn.get(parent).add(child);
                else {
                    conn.put(parent, new ArrayList<>());
                    conn.get(parent).add(child);
                }
                if(conn.containsKey(child))
                    conn.get(child).add(parent);
                else {
                    conn.put(child, new ArrayList<>());
                    conn.get(child).add(parent);
                }
                if(parentCount.containsKey(child))
                    isTree = false;
                else
                    parentCount.put(child, 1);
                ++edgeCount;
                startVertex = parent;
            }
        }
        System.out.print(output);
	}
	
	static void init() {
	    conn = new HashMap<>();
		parentCount = new HashMap<>();
		edgeCount = startVertex = 0;
		isTree = true;
	}
	
	static void dfs(HashMap<Integer, Boolean> visited, int vertex, int[] nodeCount) {
        if(!visited.containsKey(vertex)) {
            visited.put(vertex, true);
            ++nodeCount[0];
            for(Integer nextVertex : conn.get(vertex))
                dfs(visited, nextVertex, nodeCount);
        }
    }
}
